### PR TITLE
xRegistry: Fixed issue unable to create registry value with data zero. (Fixes #276)

### LINKED ISSUE #276: xRegistry: Unable to create binary registry value with data '0x00'
## Description
I want to create binary registry value with data '0x00', but xRegistry resource set none value.

## Detail
I want to create binary reg value with data zero
zero

For that, write and exec psdsc configuration like...
```PowerShell
xRegistry Zero_RegBinary
{
    Ensure = "Present"
    Key = "HKEY_LOCAL_MACHINE\SYSTEM\Tests"
    ValueName = "TestValue"
    ValueData = "0x00"    # same result if you set ValueData="0", 
    ValueType = "Binary"
}
```
Then, sadly, xRegistry resource create a reg value that has 'zero-length binary value'
none

 I confirmed the issue also have the In-box Regisrtry resource in Windows 10 1607.

## How to fix
I've discovered the issue caused by [this code](https://github.com/PowerShell/xPSDesiredStateConfiguration/blob/37043240c09220e134c76a427b2364b949c7e424/DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1#L1198).
```Powershell
$val = $Data[0].TrimStart('0x')
```
Apparently the code is intended to remove the first '0x' of $Data[0].
but if $Data[0] = '0' or '0x00', empty value is set in $val.
Perhaps this is undesired behavior.

I would like to propose to modify the code like...
```Powershell
if($Data[0].StartsWith('0x')) {
   $val = $Data[0].Substring(2)
}
else {
   $val = $Data[0]
}
```
It's a little redundant, but I think it is easy to understand and behave properly.

### PR DESCRIPTION
**Pull request by Amedama96**
Fix #276
- xRegistry
Modified the logic to delete the first '0x' of the string.

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/xpsdesiredstateconfiguration/281)

### CODE REVIEW

**Comment by Amedama96** [comment]
@kwirkykat
I have already signed CLA 2 days ago.
Should I need to do it once again ?

**Comment by kwirkykat** [comment]
@Amedama96 CLA bot is having issues. I'm going to close the PR and then re-open it to try to get it to cooperate.

**Comment by TravisEz13** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1, line 1134 at r1](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281#-KXmA__JqRDU67vPTcE1:-KXmA__JqRDU67vPTcE2:bb8m5n7) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/c048715e03fcc83a30b2acaea74edfe08aaecabc/DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1#L1134)):*
> ```PowerShell
>             {
>                 $retVal = $null
>                 if ($Data[0].StartsWith('0x'))
> ```

If we repeat this code 3 times, the code should be a function, with tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281)*

**Comment by LiamFisher92** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1, line 1136 at r1](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281#-KXpW5CKBa2JwSoqgfDW:-KXpW5CMdOL5WlkjBkKT:b-fpnt7s) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/c048715e03fcc83a30b2acaea74edfe08aaecabc/DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1#L1136)):*
> ```PowerShell
>                 if ($Data[0].StartsWith('0x'))
>                 {
>                     $val = $Data[0].Substring(2)
> ```

Sorry for the duplicate comment, I didn't realize you had a PR submitted.
I don't think the issue is as simple as taking a sub string of the string value. A value with '0x' as a prefix means that the value is a Hexadecimal value (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, A, B, C, D, E, F), so if the value contained any hexadecimal letters, [Int]::TryParse('F', [ref] $out) would fail, and you would be in the same situation. The only way I know of getting round this is to use [System.Convert]::ToInt32([string] $value, [int32] fromBase).
Binary is Base 2, Integer (denary) is base 10, and Hexadecimal is base 16, so to convert a base 16 value of '0xF' to a base 10 integer you would use:
[System.Convert]::ToInt32('0xF', 16)
This will give you the true value of the integer (15). Hope this helps, let me know if you need any more help.

---

*[DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1, line 1175 at r1](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281#-KXpWDOxckR9V5bwCTcz:-KXpWDOxckR9V5bwCTd-:b70dp9r) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/c048715e03fcc83a30b2acaea74edfe08aaecabc/DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1#L1175)):*
> ```PowerShell
>                 if ($Data[0].StartsWith('0x'))
>                 {
>                     $val = $Data[0].Substring(2)
> ```

Comment is as above.

---

*[DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1, line 1214 at r1](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281#-KXpWGSrMmn0eAGB2MKQ:-KXpWGSrMmn0eAGB2MKR:b70dp9r) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/c048715e03fcc83a30b2acaea74edfe08aaecabc/DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1#L1214)):*
> ```PowerShell
>             if ($Data[0].StartsWith('0x'))
>             {
>                 $val = $Data[0].Substring(2)
> ```

Comment is as above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281)*

**Comment by Amedama96** [comment]
Hi @LiamFisher92
I appreciate your comment !
and sorry, I think that has no problem for the logic of parsing string to bytes array.

This is the code for parsing. (prefix '0x' has removed already)
```Powershell
for ($i = 0 ; $i -lt ($val.Length-1) ; $i = $i+2)
{
    $byteArray += [System.Byte]::Parse($val.Substring($i, 2), 'HexNumber')
}
```
[System.Byte]::Parse('0', 'HexNumber') give you byte (0).
[System.Byte]::Parse('F', 'HexNumber') give you byte (15).
[System.Byte]::Parse('10', 'HexNumber') give you byte (16).

These results are correct, aren't you?
Please tell me if I am misunderstanding.

**Comment by LiamFisher92** [comment]
I beg your pardon. Int32.TryParse with [System.Globalization.NumberStyles]::HexNumber handles this.
Clearly need another coffee.

**Comment by Amedama96** [comment]
@TravisEz13 
Thanks review.
Where and How should I write the test code?
It seems that there is no place to write test code of helper function.
You say I should write unit tests in the "MSFT_xRegistryResource.Tests.ps1" something like below?
```Powershell
It 'Should create a new binary registry key value with data 0' {
    $valueName = 'TestValue'
    $valueData = '0x00'
    $valueType = 'Binary'
...
```

**Comment by Amedama96** [comment]
Sadly, I found two additional related issues.
I will try to fix these issues next weekend. please wait a few days.

### Issue A
If you specify the ValueData with prefix '0x', Test-TargetResource always return $false. even if the reg value is desired.
```Powershell
Registry Zero_RegBinary
{
    Ensure = "Present"
    Key = "HKEY_LOCAL_MACHINE\SYSTEM\Tests"
    ValueName = "TestValue"
    ValueData = "0x0F"    # No problem happend if you set ValueData="0F", 
    ValueType = "Binary"
}
```

### Issue B
When the current reg binary value is '0000'.
you specify the ValueData = '00' (not '0x00'), Test passes unexpectedly.

**Comment by Amedama96** [comment]
I have resolved @TravisEz13 review and fixed additional issues.

Did My PR  mixed with @iainbrighton 's another commit ?
I'm so sorry. I don't know how to fix it. 😓

File: Tests/Unit/MSFT_xPackageResource.Tests.ps1 [hunk 1]
```diff
@@ -433,7 +435,7 @@ try
                     $msiUrl = "$baseUrl" + "package.msi"
                     New-MockFileServer -FilePath $script:msiLocation -Https
 
-                    # Test pipe connection as testing server readiness
+                    # Test pipe connection as testing server reasdiness
```
Comment (TravisEz13) [inline Tests/Unit/MSFT_xPackageResource.Tests.ps1]: Please fix misspelling

File: DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1 [hunk 2]
```diff
@@ -1279,11 +1286,12 @@ function Set-TargetResource
                 $startInfo.Arguments += ' /log "{0}"' -f $LogPath
             }
 
-            $startInfo.Arguments += " /quiet"
+            $startInfo.Arguments += " /quiet /norestart"
```
Comment (TravisEz13) [inline DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1]: Nice catch.

File: DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1 [hunk 3]
```diff
@@ -1341,38 +1350,31 @@ function Set-TargetResource
...
-                $exitCode = 0
-
-                $process.Start() | Out-Null
-
-                # Identical to $fileExtension -eq '.exe' -and $logPath
-                if ($logStream)
+                [int] $exitCode = 0
+                if($PSBoundParameters.ContainsKey('RunAsCredential'))
                 {
-                    $process.BeginOutputReadLine()
```
Comment (TravisEz13) [inline DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1]: functional regression of the logstream functionality (but looking at the code I'm not 100% sure it worked)

File: DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1 [hunk 4]
```diff
@@ -1638,4 +1639,401 @@ function Assert-FileSignatureValid
...
+}
+
+<#
+    .SYNOPSIS
+        Registers PInvoke to run a process as a user.
+#>
+function Register-PInvoke
+{
+    $programSource = @'
+        using System;
```
Comment (TravisEz13) [inline DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1]: Did you get this code from another resource?  If so, please add a comment about where you got it from, please?

**Comment by iainbrighton** [comment]
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1, line 1289 at r4](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281#-KY2fLoVMivL1xvFBfBz-r4-1289:-KYL6VD_5z3PconpG8Sj:b2bg7sl) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/db01c6512e899e763d89b7824f70edaed5871801/DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1#L1289)):*
<details><summary><i>Previously, TravisEz13 (Travis Plunk) wrote…</i></summary><blockquote>

Nice catch.
</blockquote></details>

Thanks! Not sure how the xPackage changes ended up in this PR as they were all merged in https://github.com/PowerShell/xPSDesiredStateConfiguration/pull/278 ?

---

*[DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1, line 1727 at r4](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281#-KY2fLoVMivL1xvFBfBz-r4-1727:-KYL6s_t7KqklxE7fd4I:b-yh7m61) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/db01c6512e899e763d89b7824f70edaed5871801/DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1#L1727)):*
<details><summary><i>Previously, TravisEz13 (Travis Plunk) wrote…</i></summary><blockquote>

Did you get this code from another resource?  If so, please add a comment about where you got it from, please?
</blockquote></details>

This comes from older versions of the [xPackage resource](https://github.com/PowerShell/xPSDesiredStateConfiguration/tree/850a65d893417afb4b7f8f3f99732c29541e591b)!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281)*

**Comment by TravisEz13** [comment]
Reviewed 2 of 2 files at r1.
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1, line 1727 at r4](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281#-KY2fLoVMivL1xvFBfBz-r4-1727:-KYLTd66BbN-Joceedq7:brae6ub) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/db01c6512e899e763d89b7824f70edaed5871801/DSCResources/MSFT_xPackageResource/MSFT_xPackageResource.psm1#L1727)):*
<details><summary><i>Previously, iainbrighton (Iain Brighton) wrote…</i></summary><blockquote>

This comes from older versions of the [xPackage resource](https://github.com/PowerShell/xPSDesiredStateConfiguration/tree/850a65d893417afb4b7f8f3f99732c29541e591b)!
</blockquote></details>

Please add a comment to the code of where the code from.  And probably file an issue that we should have a common module to do this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281)*

**Comment by Amedama96** [comment]
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1, line 1134 at r1](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281#-KXmA__JqRDU67vPTcE1:-KYOobr2AZGrP92xzngG:b-xi29w9) ([raw file](https://github.com/powershell/xpsdesiredstateconfiguration/blob/c048715e03fcc83a30b2acaea74edfe08aaecabc/DSCResources/MSFT_xRegistryResource/MSFT_xRegistryResource.psm1#L1134)):*
<details><summary><i>Previously, TravisEz13 (Travis Plunk) wrote…</i></summary><blockquote>

If we repeat this code 3 times, the code should be a function, with tests.
</blockquote></details>

Added helper function 'Remove-PrefixString'

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281)*

**Comment by TravisEz13** [comment]
Reviewed 3 of 4 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xpsdesiredstateconfiguration/281)*

**Comment by Amedama96** [comment]
Those xRegistry resource issues fixed and merged #299 by @kwirkykat 
Thanks all members and 
I apologize for any inconvenience caused by lack of knowledge about github.

**Pull request closed without merging**